### PR TITLE
fixes #2567 corrects race condition in tablet metadata verification

### LINKED ISSUE #2567: Data files in Tablet differ from in-memory data
**Describe the bug**
I saw this error while running Randomwalk tests on the 2.1 main branch.
`Data files in 28;7;6 differ from in-memory data`

**Versions (OS, Maven, Java, and others, as appropriate):**
 - Affected version(s) of this project: [2.1]
 - OS: [Ubuntu]

**To Reproduce**
Steps to reproduce the behavior (or a link to an example repository that reproduces the problem):
1. Setup Uno. I had 2 tservers running locally.
2. Ingest some data for a while using Randomwalk MultiTable test

Tserver log activity:
<pre>
2022-03-15T09:17:46,688 [memory.LargestFirstMemoryManager] DEBUG: COMPACTING !0;~<  total = 203,913,354 ingestMemory = 22,659,626
2022-03-15T09:17:47,651 [memory.LargestFirstMemoryManager] DEBUG: COMPACTING 28;9;8  total = 226,570,072 ingestMemory = 45,313,436
2022-03-15T09:17:50,471 [tablet.Tablet] ERROR: Data files in 28;7;6 differ from in-memory data ...
2022-03-15T09:17:50,655 [tablet.files] DEBUG: Flushed 28;7;6 created hdfs://localhost:8020/accumulo/tables/28/t-00002pt/F00004hu.rf from [memory]
2022-03-15T09:17:50,657 [tablet.files] DEBUG: Flushed !0;~< created hdfs://localhost:8020/accumulo/tables/!0/table_info/F00004hz.rf from [memory]
2022-03-15T09:17:50,662 [tablet.files] DEBUG: Compacting !0;~< on i.meta.small for SYSTEM from [F00004hz.rf, A00004hy.rf] size 20 KB
2022-03-15T09:17:50,676 [memory.LargestFirstMemoryManager] DEBUG: COMPACTING 28;7;6  total = 272,062,261 ingestMemory = 90,808,517
2022-03-15T09:17:50,676 [memory.LargestFirstMemoryManager] DEBUG: chosenMem = 45,493,788 chosenIT = 2.31 load 45,574,937
2022-03-15T09:17:50,677 [tablet.Tablet] DEBUG: MinC initiate lock 0.00 secs
2022-03-15T09:17:51,540 [memory.LargestFirstMemoryManager] DEBUG: BEFORE compactionThreshold = 0.056 maxObserved = 339,854,955
2022-03-15T09:17:51,540 [memory.LargestFirstMemoryManager] DEBUG: AFTER compactionThreshold = 0.056
2022-03-15T09:17:51,596 [constraints.ConstraintChecker] DEBUG: Loaded constraint org.apache.accumulo.server.constraints.MetadataConstraints for !0
2022-03-15T09:17:51,611 [tablet.files] DEBUG: Flushed 28;9;8 created hdfs://localhost:8020/accumulo/tables/28/t-00002pv/F00004i0.rf from [memory]
2022-03-15T09:17:51,613 [accumulo.audit] INFO : operation: permitted; user: root; client: 127.0.0.1:50726; action: authenticate;
2022-03-15T09:17:51,652 [tablet.files] DEBUG: Compacted !0;~< for SYSTEM created hdfs://localhost:8020/accumulo/tables/!0/table_info/A00004i2.rf from [F00004hz.rf, A00004hy.rf]
</pre>

### PR DESCRIPTION
**Pull request by keith-turner**

### CODE REVIEW

**Comment by keith-turner** [comment]
The diffs for this are much easier to look at when hiding whitespace diffs.  @milleruntime I think this may fix the problem you saw in #2567, but I am not completely sure yet.  I want to come back and review it myself in a few days.  My goal in this PR was to detect changes for the not just updating of the java map of files but also changes written to the metadata table.  As you identified in the issue, the existing code only detects changes to the java map but it does not detect overlapping writes to the metadata table.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java [hunk 1]
```diff
@@ -1419,7 +1419,11 @@ private void closeConsistencyCheck() {
         }
       });
 
-      compareToDataInMemory(tabletMeta);
+      if (!tabletMeta.getFilesMap().equals(getDatafileManager().getDatafileSizes())) {
+        String msg = "Data files in " + extent + " differ from in-memory data "
+            + tabletMeta.getFilesMap() + " " + getDatafileManager().getDatafileSizes();
+        log.error(msg);
+      }
```
Comment (ctubbsii) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1426]: Could these maps change between the check and the message construction?

```suggestion
      var filesMap = tabletMeta.getFilesMap();
      var datafileSizes = getDatafileManager().getDatafileSizes();
      if (!filesMap.equals(datafileSizes)) {
        String msg = "Data files in " + extent + " differ from in-memory data " + filesMap + " " + datafileSizes;
        log.error(msg);
      }
```

If the objects themselves are mutable, then it could be even more of a problem and we might need to do a protective copy to do the comparison.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java [hunk 2]
```diff
@@ -1435,22 +1439,28 @@ private void closeConsistencyCheck() {
...
+      return;
     }
-  }
 
-  public synchronized void compareTabletInfo(Long updateCounter, TabletMetadata tabletMetadata) {
-    if (isClosed() || isClosing()) {
+    if (updateCounter.overlapsUpdate() || !updateCounter.equals(this.getUpdateCount())) {
+      // for these cases do not even bother checking if the files are the same
       return;
     }
```
Comment (ctubbsii) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1451]: :+1: for the first case, where we're checking while an update is occurring. But, I'm not sure I understand the second case. What situation does that skip?

Comment (keith-turner) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1451]: In the second case there is evidence that a change has occurred, so its an optimization to avoid comparing the maps.

Comment (ctubbsii) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1451]: > In the second case there is evidence that a change has occurred, so its an optimization to avoid comparing the maps.

Isn't this evidence that either:
1. a change has occurred, or
2. a possible inconsistency has been detected?

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java (hunk 1 above)
Comment (keith-turner) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1426]: > Could these maps change between the check and the message construction?

No, neither one of them should change.

What `tabletMeta.getFilesMap()` returns comes originally from Ample when it read from the metadata table.  If this were changing it would be a really weird bug that would have a much wider impact.  What `getDatafileManager().getDatafileSizes()` returns is a copy of its files made using the same lock this method is holding.  So it should not change given the current code, but it would probably be cleaner to only get it once.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/MetadataUpdateCount.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    this.startedCount = startedCount;
+    this.finishedCount = finishedCount;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o)
+      return true;
+    if (o == null || getClass() != o.getClass())
+      return false;
```
[the remark this answers is not in the document]
Comment (keith-turner) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/MetadataUpdateCount.java:37]: That code was generated by intellij.  I think the original code will reject subclasses where your suggestion will let subclasses pass through.  For this case I like rejecting subclasses as I would like the equals method to be very strict.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java (hunk 2 above)
Comment (keith-turner) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1451]: If a change has occurred then we can reliably check for inconsistencies, so need to rely on the fact that it will eventually check again later.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java (hunk 1 above)
Comment (keith-turner) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1426]: I made an update re this in 14cf0d2c80c5ba55d54d25b1b2fa221a6f13dee8.  In that update I also adjusted the log message.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java (hunk 2 above)
Comment (keith-turner) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1451]: The big picture is that we are periodically reading a tablets metadata from the metadata table and comparing that to what the tablet has in memory.  We only want to do this comparison when there were no changes to the tablet while were in the process of reading the metadata table.   So when we know for sure a change has occurred to the tablet while we were reading from the metadata table there is no need to compare what we got from the metadata table to the tablets current files.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java [hunk 4]
```diff
@@ -500,8 +519,8 @@ public int getNumFiles() {
     return datafileSizes.size();
   }
 
-  public long getUpdateCount() {
-    return updateCount;
+  public MetadataUpdateCount getUpdateCount() {
+    return new MetadataUpdateCount(metadataUpdatesStarted.get(), metadataUpdatesCompleted.get());
```
Comment (keith-turner) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java:523]: I have a slight concern about this code.  It may matter that the start count is read before the completed count, not sure still thinking out it.  Thinking of replacing the two separate atomic longs to a single atomic reference to a counter object with two longs.  Then I don't have to worry about this issue.

Comment (keith-turner) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java]: changed this in fd31a15

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java (hunk 2 above)
Comment (ctubbsii) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1451]: I misunderstood the object returned from `getUpdateCount()` as still being a primitive, and not the new MetadataUpdateCount type you created. It makes sense now. Thanks.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java [hunk 5]
```diff
@@ -246,22 +247,27 @@ else if (!bulkDir.equals(tpath.getTabletDir()))
       }
     }
 
-    synchronized (tablet) {
-      for (Entry<StoredTabletFile,DataFileValue> tpath : newFiles.entrySet()) {
-        if (datafileSizes.containsKey(tpath.getKey())) {
-          log.error("Adding file that is already in set {}", tpath.getKey());
+    metadataUpdateCount.updateAndGet(oldCount -> oldCount.incrementStart());
```
Comment (ctubbsii) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java:250]:
```suggestion
    metadataUpdateCount.updateAndGet(MetadataUpdateCount::incrementStart);
```

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java [hunk 6]
```diff
@@ -246,22 +247,27 @@ else if (!bulkDir.equals(tpath.getTabletDir()))
...
+        tablet.computeNumEntries();
+      }
 
-    for (Entry<StoredTabletFile,DataFileValue> entry : newFiles.entrySet()) {
-      TabletLogger.bulkImported(tablet.getExtent(), entry.getKey());
+      for (Entry<StoredTabletFile,DataFileValue> entry : newFiles.entrySet()) {
+        TabletLogger.bulkImported(tablet.getExtent(), entry.getKey());
+      }
+    } finally {
+      metadataUpdateCount.updateAndGet(oldCount -> oldCount.incrementFinish());
```
Comment (ctubbsii) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java:270]:
```suggestion
      metadataUpdateCount.updateAndGet(MetadataUpdateCount::incrementFinish);
```

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java [hunk 7]
```diff
@@ -357,35 +363,41 @@ else if (!bulkDir.equals(tpath.getTabletDir()))
       tablet.finishClearingUnusedLogs();
     }
 
-    do {
-      try {
-        // the purpose of making this update use the new commit session, instead of the old one
-        // passed in, is because the new one will reference the logs used by current memory...
+    metadataUpdateCount.updateAndGet(oldCount -> oldCount.incrementStart());
```
Comment (ctubbsii) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java:366]:
```suggestion
    metadataUpdateCount.updateAndGet(MetadataUpdateCount::incrementStart);
```

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java [hunk 8]
```diff
@@ -357,35 +363,41 @@ else if (!bulkDir.equals(tpath.getTabletDir()))
...
-      }
 
-      tablet.flushComplete(flushId);
+        tablet.flushComplete(flushId);
 
-      t2 = System.currentTimeMillis();
+        t2 = System.currentTimeMillis();
+      }
+    } finally {
+      metadataUpdateCount.updateAndGet(oldCount -> oldCount.incrementFinish());
```
Comment (ctubbsii) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java:400]:
```suggestion
      metadataUpdateCount.updateAndGet(MetadataUpdateCount::incrementFinish);
```

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java [hunk 9]
```diff
@@ -431,49 +443,56 @@ StoredTabletFile bringMajorCompactionOnline(Set<StoredTabletFile> oldDatafiles,
...
-    synchronized (tablet) {
-      t1 = System.currentTimeMillis();
-
-      Preconditions.checkState(datafileSizes.keySet().containsAll(oldDatafiles),
-          "Compacted files %s are not a subset of tablet files %s", oldDatafiles,
-          datafileSizes.keySet());
-      if (dfv.getNumEntries() > 0) {
-        Preconditions.checkState(!datafileSizes.containsKey(newFile),
-            "New compaction file %s already exist in tablet files %s", newFile,
+    metadataUpdateCount.updateAndGet(oldCount -> oldCount.incrementStart());
```
Comment (ctubbsii) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java:446]:
```suggestion
    metadataUpdateCount.updateAndGet(MetadataUpdateCount::incrementStart);
```

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java [hunk 10]
```diff
@@ -431,49 +443,56 @@ StoredTabletFile bringMajorCompactionOnline(Set<StoredTabletFile> oldDatafiles,
...
-    if (!filesInUseByScans.isEmpty())
-      log.debug("Adding scan refs to metadata {} {}", extent, filesInUseByScans);
-    ManagerMetadataUtil.replaceDatafiles(tablet.getContext(), extent, oldDatafiles,
-        filesInUseByScans, newFile, compactionIdToWrite, dfv,
-        tablet.getTabletServer().getClientAddressString(), lastLocation,
-        tablet.getTabletServer().getLock(), ecid);
-    tablet.setLastCompactionID(compactionIdToWrite);
-    removeFilesAfterScan(filesInUseByScans);
+    } finally {
+      metadataUpdateCount.updateAndGet(oldCount -> oldCount.incrementFinish());
```
Comment (ctubbsii) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java:494]:
```suggestion
      metadataUpdateCount.updateAndGet(MetadataUpdateCount::incrementFinish);
```

**Review by ctubbsii — CHANGES_REQUESTED** [review]
LGTM, except these lambdas.

**Review by ctubbsii — APPROVED** [review]

**Comment by keith-turner** [comment]
Before merging this thinking next week of adding some trace level logging of the counts and running this to see if things are really working as intended.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java [hunk 11]
```diff
@@ -1435,23 +1439,34 @@ private void closeConsistencyCheck() {
...
-    if (!tabletMetadata.getFilesMap().equals(getDatafileManager().getDatafileSizes())) {
-      String msg = "Data files in " + extent + " differ from in-memory data "
-          + tabletMetadata.getFilesMap() + " " + getDatafileManager().getDatafileSizes();
-      log.error(msg);
-    }
-  }
-
-  public synchronized void compareTabletInfo(Long updateCounter, TabletMetadata tabletMetadata) {
+  public synchronized void compareTabletInfo(MetadataUpdateCount updateCounter,
+      TabletMetadata tabletMetadata) {
```
Comment (milleruntime) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1443]: Does it matter which thread (the main thread vs during a close) is doing the check? This was one thing I was wondering while doing testing. If we add a boolean here of who is calling the check, we could bail if it is not the correct thread.

Comment (keith-turner) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1443]: It should not matter which thread is doing the check. The main thing that matters is the order in which the two parameters are that are passed to the method are acquired.  The updateCounter must be acquired before tabletMetadata by the code calling this method (and they must go with the same tablet).

Comment (keith-turner) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1443]: Maybe we could add a key extent to the MetadataUpdateCount to allow more validations (make sure counts are for expected extent) in the method.

Comment (keith-turner) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1454]: I added some extra validations in e2580d9.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java [hunk 12]
```diff
@@ -69,13 +70,16 @@
   // ensure we only have one reader/writer of our bulk file notes at at time
   private final Object bulkFileImportLock = new Object();
 
-  // This must be incremented whenever datafileSizes is mutated
-  private long updateCount;
+  // This must be incremented before and after datafileSizes and metadata table updates. These
+  // counts allow detection of overlapping operation w/o placing a lock around metadata table
```
Comment (milleruntime) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java:74]: I _think_ I know what you are trying to say with "overlapping operation" but maybe you could expand on it with an additional sentence? Or maybe reword "overlapping operation" to something else?

**Review by milleruntime — COMMENTED** [review]
I think it looks good. But I am just trying to understand what is going on so don't wait for me.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java [hunk 13]
```diff
@@ -1435,23 +1439,56 @@ private void closeConsistencyCheck() {
...
+      // table) after the equality check above then we know the tablet did not do any updates while
+      // we were reading metadata and then comparing.
+      var latestCount = this.getUpdateCount();
+      if (updateCounter.overlapsUpdate() || !updateCounter.equals(latestCount)) {
+        log.trace(
+            "AMCC Tablet {} may have been updating its metadata while it was being read for check, so skipping check {} {}",
+            tabletMetadata.getExtent(), updateCounter, latestCount);
+      } else {
+        log.error("AMCC Data files in {} differ from in-memory data {} {} {} {}", extent,
+            tabletMetadata.getFilesMap(), dataFileSizes, updateCounter, latestCount);
```
Comment (milleruntime) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1486]:
```suggestion
        log.error(
            "AMCC Data files in {} differ from in-memory data {}. dataFileSize: {} updateCounter: {} latestCount: {}",
            extent, tabletMetadata.getFilesMap(), dataFileSizes, updateCounter, latestCount);
```

Comment (milleruntime) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1486]: I am not sure what "AMCC" stands for but I guess it's OK to add uniqueness to the log statement. I just don't want to confuse users.

Comment (keith-turner) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1486]: AMCC was for Accumulo Metadata Consistency Check.  I can add a comment for that.  Do you think it should be removed?  I added because I like to be able to easily grep for all related messages.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java (hunk 12 above)
Comment (keith-turner) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java]: I made multiple comment updates in 7d6b61e

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/MetadataUpdateCount.java [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.tserver.tablet;
+
+import java.util.Objects;
+
+import org.apache.accumulo.core.dataImpl.KeyExtent;
+
+public class MetadataUpdateCount {
```
[the remark this answers is not in the document]
Comment (keith-turner) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/MetadataUpdateCount.java:33]: Took a stab at this in 7d6b61e

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java (hunk 12 above)
Comment (milleruntime) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java]: Thanks, those are a BIG help!

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java (hunk 13 above)
Comment (milleruntime) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1486]: I like the ability to grep as well. I think a comment above where it gets logged would be nice. If a user is curious and they go to the code then they will see what it means.

Comment (keith-turner) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1490]: Added a comment in e60b6bb9b99a0783b4dbcd3aadb882123efed04c

Comment (ctubbsii) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1491]: Jargon in the logs is very confusing. People won't know to grep for this specifically, and even if they go to the code to figure it out, it doesn't seem like it should be necessary. I'm not a fan of non-obvious jargon in log messages. It's also just as easy to grep for `'Data files in.*differ from in-memory'` or similar, if you know what you're expecting. If the jargon is still desired, I think it would be better to spell it out explicitly, even if it makes the log message longer.

Comment (ctubbsii) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1491]: Maybe it could be put in a class called "AccumuloMetadataConsistencyCheck" or just use a named logger with a name like that.

Comment (milleruntime) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1491]: I think its fine for logging if it was shorthand for some type of name. I wouldn't consider it jargon if it was an acronym or an abbreviation of something.

Comment (keith-turner) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java]: > Maybe it could be put in a class called "AccumuloMetadataConsistencyCheck" or just use a named logger with a name like that.

I don't want to refactor the code like that just for the log messages.  I removed the AMCC prefix from the error message and left it on the trace messages in 2296b47.  My thinking is that the error message is unsolicited, and the prefix may be confusing, however when someone is turning on tracing they are soliciting as much information as possible and I think this extra information may be useful (can be used for selection, deselection, tabulation, etc) even if a bit confusing at first.

Comment (ctubbsii) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java]: Seems like a reasonable compromise.

**Comment by ctubbsii** [comment]
@keith-turner wrote:
> @ctubbsii any interest in looking at it again before I merge?

No, it was fine before. I skimmed, and didn't see anything substantively different from what I previously reviewed. So, unless there's something specific that you changed that you think might warrant a closer look, I'm fine with merging (the merge conflict will have to be resolved... should be easy... I think it's just indentation conflicts on most lines).
